libethereum/EthereumPeer.cpp

```diff
@@ -246,22 +248,77 @@ bool EthereumPeer::interpret(unsigned _id, RLP const& _r)
 		host()->onPeerTransactions(dynamic_pointer_cast<EthereumPeer>(dynamic_pointer_cast<EthereumPeer>(shared_from_this())), _r);
 		break;
 	}
-	case GetBlockHashesPacket:
+	case GetBlockHeadersPacket:
 	{
-		h256 later = _r[0].toHash<h256>();
-		unsigned limit = _r[1].toInt<unsigned>();
-		clog(NetMessageSummary) << "GetBlockHashes (" << limit << "entries," << later << ")";
-		unsigned c = min<unsigned>(host()->chain().number(later), min(limit, c_maxHashesToSend));
+		/// Packet layout:
+		/// [ block: { P , B_32 }, maxHeaders: P, skip: P, reverse: P in { 0 , 1 } ]
+		auto blockId = _r[0];
+		auto maxHeaders = _r[1].toInt<bigint>();
+		auto skip = _r[2].toInt<bigint>();
+		auto reverse = _r[3].toInt<bool>();
+
+		auto& bc = host()->chain();
+
+		auto numHeadersToSend = maxHeaders <= c_maxHeadersToSend ? static_cast<unsigned>(maxHeaders) : c_maxHeadersToSend;
+
+		if (skip > c_maxBlockSkip)
+		{
+			clog(NetAllDetail) << "Requested block skip is too big: " << skip << ", max: " << c_maxBlockSkip;
+			break;
+		}
+
+		auto step = static_cast<unsigned>(skip) + 1;
+
+		h256 blockHash;
+		if (blockId.size() == 32) // block id is a hash
+		{
+			blockHash = blockId.toHash<h256>();
+			//blockNumber = host()->chain().number(blockHash);
+			clog(NetMessageSummary) << "GetBlockHeaders (block (hash): " << blockHash
+			<< ", maxHeaders: " << maxHeaders
+			<< ", skip: " << skip << ", reverse: " << reverse << ")";
+
+			if (!reverse)
+			{
+				auto n = bc.number(blockHash);
+				n += step * numHeadersToSend;
+				blockHash = bc.numberHash(n);
+			}
+		}
+		else // block id is a number
+		{
+			auto n = blockId.toInt<bigint>();
+			clog(NetMessageSummary) << "GetBlockHeaders (block (number): " << n
+			<< ", maxHeaders: " << maxHeaders
+			<< ", skip: " << skip << ", reverse: " << reverse << ")";
+
+			if (reverse)
```

Comment:
We cannot traverse forward (`!reverse`), so I'm looking for top requested block.